### PR TITLE
[flight recorder] record process group configuration

### PR DESCRIPTION
**Pull request by shengbao-zheng**
Summary: Record process group configuration (i.e. ranks involved in a process group) to facilitate NCCL related debugging.

Differential Revision: D53792087

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

### CODE REVIEW

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 1]
```diff
@@ -1500,6 +1501,15 @@ const int& ProcessGroupNCCL::globalRank() const {
   return globalRank;
 }
 
+const std::vector<uint64_t>& ProcessGroupNCCL::groupRanks() const {
+  if (options_->global_ranks_in_group.empty()) {
+    static std::vector<uint64_t> globalRanks(size_);
+    std::iota(globalRanks.begin(), globalRanks.end(), 0);
```
Comment (shuqiangzhang) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1507]: Before this, can we verify this is the default PG by checking if uid_ == 0?

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp [hunk 2]
```diff
@@ -724,6 +724,9 @@ class TORCH_API ProcessGroupNCCL : public Backend {
   // return the rank_ of the the very first PG created, aka, default global PG.
   const int& globalRank() const;
 
+  // Returns the global ranks of of a PG.
```
Comment (shuqiangzhang) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:727]: nit, 2 'of'

File: torch/csrc/distributed/c10d/TraceUtils.h [hunk 3]
```diff
@@ -475,6 +477,18 @@ struct NCCLTraceBuffer {
     return id_++;
   }
 
+  void record_pg(size_t pg_id, std::vector<uint64_t> ranks) {
+    if (!enabled_) {
+      return;
+    }
+    std::lock_guard<std::mutex> guard(mutex_);
+    if (pg_id_to_ranks_.find(pg_id) != pg_id_to_ranks_.end()) {
```
Comment (shuqiangzhang) [inline torch/csrc/distributed/c10d/TraceUtils.h:485]: When will this happen? or we just need to update the map whenever we call record_pg by pg_id_to_ranks_[pg_id] = ranks;

File: torch/csrc/distributed/c10d/TraceUtils.h [hunk 4]
```diff
@@ -475,6 +477,18 @@ struct NCCLTraceBuffer {
     return id_++;
   }
 
+  void record_pg(size_t pg_id, std::vector<uint64_t> ranks) {
```
Comment (shuqiangzhang) [inline torch/csrc/distributed/c10d/TraceUtils.h:480]: nit: more specific, record_pg_ranks?

File: torch/csrc/distributed/c10d/TraceUtils.h [hunk 5]
```diff
@@ -660,10 +675,19 @@ struct NCCLTraceBuffer {
       dict.insert(frames_key, frames);
       entries.push_back(dict);
     }
+    auto pg_config = new_dict();
+    for (auto it = pg_id_to_ranks_.begin(); it != pg_id_to_ranks_.end(); it++) {
+      auto ranks = new_list();
+      for (uint64_t rank : it->second) {
```
Comment (shuqiangzhang) [inline torch/csrc/distributed/c10d/TraceUtils.h:681]: nit, const auto&

File: torch/csrc/distributed/c10d/TraceUtils.h [hunk 6]
```diff
@@ -660,10 +675,19 @@ struct NCCLTraceBuffer {
       dict.insert(frames_key, frames);
       entries.push_back(dict);
     }
+    auto pg_config = new_dict();
+    for (auto it = pg_id_to_ranks_.begin(); it != pg_id_to_ranks_.end(); it++) {
```
Comment (shuqiangzhang) [inline torch/csrc/distributed/c10d/TraceUtils.h:679]: nit: `for (const auto& [pg_id, ranks] :  pg_id_to_ranks)`

File: torch/csrc/distributed/c10d/TraceUtils.h (hunk 4 above)
Comment (shengbao-zheng) [reply torch/csrc/distributed/c10d/TraceUtils.h:480]: So far I think we only need the ranks involved in a PG so we can use record_pg_ranks for now. If more PG info is needed in the future, we can extend it and rethink the new name.

Comment (wconstab) [reply torch/csrc/distributed/c10d/TraceUtils.h:480]: Does it make sense to include the nccl comm id in this too?

File: torch/csrc/distributed/c10d/TraceUtils.h [hunk 7]
```diff
@@ -573,6 +587,7 @@ struct NCCLTraceBuffer {
     // Update whenever changing contents or formatting of the dump
     // (minor when adding fields, major when changing existing fields)
     c10::IValue version_val = "1.1";
```
Comment (wconstab) [inline torch/csrc/distributed/c10d/TraceUtils.h:589]: It should be 1.2 after adding the field

File: torch/csrc/distributed/c10d/TraceUtils.h (hunk 4 above)
Comment (shengbao-zheng) [reply torch/csrc/distributed/c10d/TraceUtils.h:480]: I think we can revisit this (adding nccl comm id) if the current info is not enough?

**Review by shuqiangzhang — APPROVED** [review]

**Pull request closed without merging**